### PR TITLE
gpstart: test and refactor #8956

### PR DESCRIPTION
**Pull request by jchampio**
Follow-up to PR #8956, in preparation for backport to 5 and 6. The general roadmap:
- Prefactor an existing behave step for use on clusters other than a demo cluster.
- Add regression tests for gpstart.
- Further simplify the previously-buggy implementation now that there's less code to obfuscate the meaning. (This commit isn't planned for backport since it doesn't introduce a difference in behavior.)

## Checklist
- [x] Add tests for the change
- [ ] ~Document changes~
- [ ] ~Communicate in the mailing list if needed~
- [x] Pass `make installcheck` ([pipeline](https://cm.ci.gpdb.pivotal.io/teams/main/pipelines/invalid-segments))
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: gpMgmt/test/behave/mgmt_utils/steps/replication_slots_utils.py [hunk 1]
```diff
@@ -79,9 +79,24 @@ def step_impl(context):
     stop_database(context)
 
 
+@given(u'a mirror has crashed')
 @when(u'a mirror has crashed')
```
Comment (jchampio) [inline gpMgmt/test/behave/mgmt_utils/steps/replication_slots_utils.py]: @berlin-ab If you have time, I'd like to know if you have any input on the changes we made to this step.

File: gpMgmt/bin/gpstart [hunk 2]
```diff
@@ -449,21 +449,11 @@ class GpStart:
         # segments marked down
         invalid_segs = self.gparray.get_invalid_segdbs()
 
-        # now produce the list of segments to actually start
-        dbIdsToNotStart = {}
-        for segment in invalid_segs:
-            dbIdsToNotStart[segment.getSegmentDbId()] = True
-
         for seg in invalid_segs:
```
Comment (kalensk) [inline gpMgmt/bin/gpstart]: I am not sure what the purpose of this original code was? It seems like `dbIdsToNotStart` was never being returned or really used?! I guess I am a little confused at understanding the original code in order to help review this commit.

File: gpMgmt/test/behave/mgmt_utils/gpstart.feature [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    @concourse_cluster
+    @demo_cluster
+    Scenario: gpstart correctly identifies down segments
+        Given the database is running
+          And a mirror has crashed
+          And the database is not running
+         When the user runs "gpstart -a"
+         Then gpstart should return a return code of 0
+          And gpstart should print "Skipping startup of segment marked down in configuration" to stdout
+          And gpstart should print "Skipped segment starts \(segments are marked down in configuration\) += 1" to stdout
```
Comment (kalensk) [inline gpMgmt/test/behave/mgmt_utils/gpstart.feature]: What is `+= 1`?

File: gpMgmt/bin/gppylib/test/unit/test_unit_gpstart.py [hunk 4]
```diff
@@ -181,6 +181,30 @@ def test_standby_startup_skipped(self):
         messages = [msg[0][0] for msg in self.subject.logger.info.call_args_list]
         self.assertIn("No standby master configured.  skipping...", messages)
 
+    def test_prepare_segment_start_correctly_partitions_segments(self):
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gppylib/test/unit/test_unit_gpstart.py]: nit: maybe name this function `test_prepare_segment_start_returns_up_and_down_segments`.

File: gpMgmt/bin/gpstart (hunk 2 above)
Comment (jchampio) [reply gpMgmt/bin/gpstart]: > I am not sure what the purpose of this original code was in this function?

You'll have to go [a while back](https://github.com/greenplum-db/gpdb/blob/762dd10696/gpMgmt/bin/gpstart#L481).

> I guess I am a little confused at understanding the original code in order to help review this commit.

(Regardless of the purpose, if our refactoring introduces a behavior change, it's a bug.)

File: gpMgmt/test/behave/mgmt_utils/gpstart.feature (hunk 3 above)
Comment (jchampio) [reply gpMgmt/test/behave/mgmt_utils/gpstart.feature]: It's a regex. [space][plus]

File: gpMgmt/bin/gppylib/test/unit/test_unit_gpstart.py [hunk 5]
```diff
@@ -181,6 +181,30 @@ def test_standby_startup_skipped(self):
...
+        # Down segments
+        primary0 = Segment.initFromString("2|0|p|p|n|d|sdw1|sdw1|40000|/data/primary0")
+        mirror1 = Segment.initFromString("5|1|m|m|n|d|sdw1|sdw1|50001|/data/mirror1")
+        standby = Segment.initFromString("6|-1|m|m|n|d|sdw3|sdw3|5433|/data/standby")
+
+        gpstart.gparray = GpArray([master, primary0, primary1, mirror0, mirror1, standby])
+
+        up, down = gpstart._prepare_segment_start()
+
+        # The master and standby should not be accounted for in these lists.
```
Comment (kalensk) [inline gpMgmt/bin/gppylib/test/unit/test_unit_gpstart.py]: I think this comment would be better if it explained "why" the master and standby are not accounted for when calling `_prepare_segment_start`. Or perhaps an even more robust idea is to have another test that demonstrates this behavior, thus having the tests document the expected behavior of the code.

**Comment by kalensk** [comment]
Friendly reminder to re-fly the master pipeline after merging.

File: gpMgmt/bin/gpstart (hunk 2 above)
Comment (kalensk) [reply gpMgmt/bin/gpstart]: FWIW I do not observe any behavior change in that commit.

File: gpMgmt/test/behave/mgmt_utils/gpstart.feature (hunk 3 above)
Comment (kalensk) [reply gpMgmt/test/behave/mgmt_utils/gpstart.feature]: I see, [space][plus] is the regex and `= 1` is part of the search string.

File: gpMgmt/bin/gppylib/test/unit/test_unit_gpstart.py (hunk 5 above)
Comment (jchampio) [reply gpMgmt/bin/gppylib/test/unit/test_unit_gpstart.py]: This is a regression (or, if you prefer, [characterization](https://en.wikipedia.org/wiki/Characterization_test)) test for a private implementation detail, so IMO the "why" is as mundane as "the current implementation requires it to be that way". It's not really important why it works that way if our real goal for the test is to pin down the behavior for refactoring.

File: gpMgmt/bin/gpstart (hunk 2 above)
Comment (dkrieger-pvtl) [reply gpMgmt/bin/gpstart]: If you start at this function `_prepare_segment_start` at the originating PR referenced #8956, you'll notice the original bug filtered out all invalid segments.  #8956 fixed that.  This PR further removes unnecessary code that similarly(IMHO) was quite confusing, and served only to add no to almost no value to the code.   The code proposed in this PR cleans up the code removed here.

File: gpMgmt/bin/gppylib/test/unit/test_unit_gpstart.py (hunk 4 above)
Comment (jchampio) [reply gpMgmt/bin/gppylib/test/unit/test_unit_gpstart.py]: Done.

**Comment by dkrieger-pvtl** [comment]
Overall, I really like this PR...it cleans up a lot of our ad-hoc usage of exectest.  And, given our frequent usage of external programs in gpupgrade, it is an important feature.

I am on board with the separate test package for execttest tests, with the common Success/Failure functions and the use of SetExecCommand helper in all tests.

The value of exectest.Select is not yet clear to me.  I mean that literally--I am not sure what exactly it saves for us.  I know there are uses of it in the code, but perhaps a before/after of the current PR but without exectest.Select would help me understand this.

**Comment by dkrieger-pvtl** [comment]
aaaaaaaarh....bad bad bad "Close and comment" button.

**Comment by jchampio** [comment]
(@dkrieger-pvtl did you mean to leave your comment on https://github.com/greenplum-db/gpupgrade/pull/160?)

**Comment by dkrieger-pvtl** [comment]
> (@dkrieger-pvtl did you mean to leave your comment on [greenplum-db/gpupgrade#160](https://github.com/greenplum-db/gpupgrade/pull/160)?)

Yes...I deleted it from this PR and added it to the correct one.

**Pull request merged**